File: packages/api-client/src/contracts/registration.ts
/** app/schemas/registration.py — RegistrationCreate / RegistrationRead (DM-S2, D-65). */
import type { SystemLifecycleStage } from "./enums";
import type { SystemDetail } from "./system";
import type { UseCaseRead } from "./use-case";
import type { ClassificationRead } from "./classification";

/** POST /v1/registrations — system-stable + first use-case facts in one body. */
export interface RegistrationCreate {
  // system-stable
  name: string;
  is_custom: boolean;
  catalogue_product_id: string | null;
  operator_role_id: string | null;
  hosting_model_id: string | null;
  lifecycle_stage: SystemLifecycleStage | null;
  owner_user_id: string | null;
  purpose: string | null;
  // first use case
  title: string;
  intended_use_category_id: string | null;
  context_blob: Record<string, unknown>;
  usage_context_id: string | null;
  human_oversight_type_id: string | null;
  data_category_ids: string[];
  affected_party_ids: string[];
  /** When present, the draft is deleted in the same transaction (D-66/SV-3). */
  draft_id?: string | null;
  /** Transient disposition signal (D-74, INV-83). Bare intake keys (e.g.
   * "operator_role_id") for explicitly confirmed derived fields; "fact:<key>"
   * for confirmed catalogue facts. Not persisted in draft_blob (B1). */
  confirmed_fields?: string[];
}

/** Response from POST /v1/registrations — full system + use case + classification,
 * no post-create refetch needed (A2). */
export interface RegistrationRead {
  system: SystemDetail;
  use_case: UseCaseRead;
  classification: ClassificationRead;
}

File: app/schemas/registration.py
"""Schemas for the atomic registration endpoint (DM-S2)."""

from __future__ import annotations

import uuid

from pydantic import BaseModel, Field

from app.models.base import SystemLifecycleStage
from app.schemas.system import SystemDetail
from app.schemas.use_cases import ClassificationRead, UseCaseRead


class RegistrationCreate(BaseModel):
    # ---- system-stable ----
    name: str = Field(min_length=1, max_length=255)
    is_custom: bool = False
    catalogue_product_id: uuid.UUID | None = None
    operator_role_id: uuid.UUID | None = None
    hosting_model_id: uuid.UUID | None = None
    lifecycle_stage: SystemLifecycleStage | None = None
    owner_user_id: uuid.UUID | None = None
    purpose: str | None = None
    # ---- first use case ----
    title: str = Field(min_length=1, max_length=255)
    use_case_purpose: str | None = None
    context_blob: dict = Field(default_factory=dict)
    usage_context_id: uuid.UUID | None = None
    human_oversight_type_id: uuid.UUID | None = None
    data_category_ids: list[uuid.UUID] = Field(default_factory=list)
    affected_party_ids: list[uuid.UUID] = Field(default_factory=list)
    # When present, the draft is deleted in the same transaction (D-66/SV-3).
    draft_id: uuid.UUID | None = None


class RegistrationRead(BaseModel):
    system: SystemDetail
    use_case: UseCaseRead
    classification: ClassificationRead
